### PR TITLE
libreddit: init at 0.10.1

### LINKED ISSUE #113066: Libreddit as service (Rust-based frontend to reddit)
**Project description**
10 second pitch: Libreddit is a portmanteau of "libre" (meaning freedom) and "Reddit". It is a private front-end like Invidious but for Reddit. Browse the coldest takes of r/unpopularopinion without being tracked.

Demo: https://libredd.it/

**Metadata**

* homepage URL: https://libredd.it/
* source URL: https://github.com/spikecodes/libreddit
* license: agpl3
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by fabaff**
###### Motivation for this change
Private front-end for Reddit

https://github.com/spikecodes/libreddit

Fixes #113066

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113185` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>libreddit: log <a href="https://termbin.com/dms56">https://termbin.com/dms56</a></li>
  </ul>
</details>

darwin patch:

```
diff --git a/pkgs/servers/libreddit/default.nix b/pkgs/servers/libreddit/default.nix
index a76c6a6de1f..7957d7ad4a2 100644
--- a/pkgs/servers/libreddit/default.nix
+++ b/pkgs/servers/libreddit/default.nix
@@ -1,6 +1,8 @@
 { lib
+, stdenv
 , rustPlatform
 , fetchFromGitHub
+, Security
 }:
 
 rustPlatform.buildRustPackage rec {
@@ -16,6 +18,8 @@ rustPlatform.buildRustPackage rec {
 
   cargoSha256 = "00m6npb9rahzk6nmhp10cn2ia3qbimvfiybk2d0x3d8wf21w84rn";
 
+  buildInputs = lib.optional stdenv.isDarwin Security;
+
   meta = with lib; {
     description = "Private front-end for Reddit";
     homepage = "https://github.com/spikecodes/libreddit";
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index b748e2eb7ec..1a4d38dd788 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -6009,7 +6009,9 @@ in
 
   libzmf = callPackage ../development/libraries/libzmf {};
 
-  libreddit = callPackage ../servers/libreddit { };
+  libreddit = callPackage ../servers/libreddit {
+    inherit (darwin.apple_sdk.frameworks) Security;
+  };
 
   libreswan = callPackage ../tools/networking/libreswan { };
 
```

**Comment by doronbehar** [comment]
@fabaff please commit @SuperSandro2000's diff and resolve merge conflicts.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

File: pkgs/servers/libreddit/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{ lib
+, stdenv
```
Comment (doronbehar) [inline pkgs/servers/libreddit/default.nix:2]:
```suggestion
, stdenv
, nixosTests
```

File: pkgs/servers/libreddit/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  src = fetchFromGitHub {
+    owner = "spikecodes";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "05qs05546b4zz60752xdvvjhihbspifjfs96lvmpw26521n3374h";
+  };
+
+  cargoSha256 = "0gk92bpqp462yrmnlkff311i6rgww4a9pan4xr435q2clajq79lw";
+
+  buildInputs = lib.optional stdenv.isDarwin Security;
```
Comment (doronbehar) [inline pkgs/servers/libreddit/default.nix:21]:
```suggestion
  buildInputs = lib.optional stdenv.isDarwin Security;
  
  passthru = {
    inherit (nixosTests) libreddit;
  };
```

**Review by doronbehar — COMMENTED** [review]
This should work I think, can be squashed into the last commit, and should make ofborg run it automatically upon every update to the package, and in this PR as well.

File: pkgs/servers/libreddit/default.nix (hunk 1 above)
Comment (fabaff) [reply pkgs/servers/libreddit/default.nix:2]: Added, thanks.

File: pkgs/servers/libreddit/default.nix (hunk 2 above)
Comment (fabaff) [reply pkgs/servers/libreddit/default.nix:22]: Thanks for the feedback.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

File: pkgs/servers/libreddit/default.nix (hunk 2 above)
Comment (doronbehar) [reply pkgs/servers/libreddit/default.nix:22]: I'm sorry but that suggestion wasn't correct and ofborg didn't run it. Sorry for the inconvenience. It should be:

```
passthru.tests = {
 inherit (nixosTests) libreddit;
};
```

Found it defined as so in other packages (`git grep nixosTests pkgs`).

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
There's a hash mismatch in the module test build:

https://logs.nix.ci/?attempt_id=5104d8ba-e859-4b1c-bb74-3426ed151b4d&key=nixos%2Fnixpkgs.113185

I verified it myself as well with `nixpkgs-review` on my machine.

File: pkgs/servers/libreddit/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+, stdenv
+, nixosTests
+, rustPlatform
+, fetchFromGitHub
+, Security
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "libreddit";
+  version = "0.3.0";
```
Comment (erdnaxe) [inline pkgs/servers/libreddit/default.nix:11]: libreddit at version 0.3.0 is running, but cannot fetch Reddit: `https://www.reddit.com/r/popular/hot.json?&raw_json=1 - Couldn't send request to Reddit: Head byte length should be less than 8kb`. The solution is to update the package.

```suggestion
  version = "0.10.1";
```

This version does no longer provide a `Cargo.lock`, so a patch seems required.

```diff
-  cargoSha256 = "0gk92bpqp462yrmnlkff311i6rgww4a9pan4xr435q2clajq79lw";
+  cargoSha256 = "0jx9sgdvp1x3hjrk1mqp52rfakmfgs5fgxz1hz4a3jj4nyrlx6k7";
+
+  cargoPatches = [
+    # a patch file to add/update Cargo.lock in the source code
+    ./add-Cargo.lock.patch
+  ];
```

With `add-Cargo.lock.patch` being <https://gist.github.com/erdnaxe/c4e25376e78f7996391d7eacccf939c1>.

I tested the module on NixOS and I am enthusiastic about this being merged in nixpkgs!

**Comment by doronbehar** [comment]
@fabaff this PR is almost perfect. Could you please update it to [0.10.1](https://github.com/spikecodes/libreddit/releases/tag/v0.10.1) and check if you still get that hash mismatch error in the nixos test? Thanks for your contribution.

File: pkgs/servers/libreddit/default.nix (hunk 3 above)
Comment (fabaff) [reply pkgs/servers/libreddit/default.nix]: Set to 0.10.1 and added your changes. Thanks a lot.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

File: nixos/modules/services/misc/libreddit.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          DynamicUser = true;
+          ExecStart = "${pkgs.libreddit}/bin/libreddit ${args}";
+          AmbientCapabilities = lib.mkIf (cfg.port < 1024) [ "CAP_NET_BIND_SERVICE" ];
+          Restart = "on-failure";
+          RestartSec = "2s";
+        };
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = [ cfg.port ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/libreddit.nix:63]: Why would I want to open that port by default?

File: nixos/modules/services/misc/libreddit.nix [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        default = "0.0.0.0";
+        example = "127.0.0.1";
+        type =  types.str;
+        description = "The address to listen on";
+      };
+
+      port = mkOption {
+        default = 8080;
+        example = 8000;
+        type = types.ints.u16;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/libreddit.nix:29]:
```suggestion
        type = types.port;
```

File: nixos/tests/libreddit.nix [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+with lib;
+
+{
+  name = "libreddit";
+  meta.maintainers = with maintainers; [ fab ];
+
+  nodes.machine =
+    { pkgs, ... }:
+    { services.libreddit.enable = true; };
```
[the remark this answers is not in the document]
Comment (doronbehar) [reply nixos/tests/libreddit.nix:11]: > I don't think we need a VM for this check.

Is there another way to spawn the service and check that it's running?

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/tests/libreddit.nix:11]: Can we start it in the checkPhase easily?

[the remark this answers is not in the document]
Comment (doronbehar) [reply nixos/tests/libreddit.nix:11]: > Can we start it in the checkPhase easily?

I think the build sandbox doesn't allow to open any ports.

**Review by doronbehar — APPROVED** [review]

File: nixos/tests/libreddit.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/tests/libreddit.nix:11]: But why do we need to open ports? we could just connect to it on localhost.

[the remark this answers is not in the document]
Comment (doronbehar) [reply nixos/tests/libreddit.nix:11]: I tried to test whether this is possible via this test:

```diff
diff --git i/pkgs/development/python-modules/fn/default.nix w/pkgs/development/python-modules/fn/default.nix
index f64d1a3d122..58daa823225 100644
--- i/pkgs/development/python-modules/fn/default.nix
+++ w/pkgs/development/python-modules/fn/default.nix
@@ -1,8 +1,13 @@
-{ lib, buildPythonPackage, fetchPypi }:
+{ lib, buildPythonPackage, fetchPypi, requests, curl }:
 
 buildPythonPackage rec {
   pname = "fn";
   version = "0.4.3";
+  postInstall = ''
+    python -m http.server 8000 &
+    curl http://0.0.0.0:8000
+  '';
+  nativeBuildInputs = [ requests curl ];
 
   meta = {
     description = ''
diff --git i/pkgs/top-level/python-packages.nix w/pkgs/top-level/python-packages.nix
index 04f95fb3d64..c2cf5556d97 100644
--- i/pkgs/top-level/python-packages.nix
+++ w/pkgs/top-level/python-packages.nix
@@ -2537,7 +2537,7 @@ in {
 
   flux-led = callPackage ../development/python-modules/flux-led { };
 
-  fn = callPackage ../development/python-modules/fn { };
+  fn = callPackage ../development/python-modules/fn { inherit (pkgs) curl; };
 
   fnvhash = callPackage ../development/python-modules/fnvhash { };
 
```

Surprisingly the command launching the http server didn't fail, but from some reason `curl` failed.. 

It's easier in general I think to maintain the test via a dedicated nixos test. In the future maintainers might want to test more things there and it'd be less comfortable to put it in the `checkPhase` - it'd require rebuilding most of the package just for changing / improving the test.

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/tests/libreddit.nix:11]: VM tests are expensive to run and should only be used when really necessary.

[the remark this answers is not in the document]
Comment (doronbehar) [reply nixos/tests/libreddit.nix:11]: > VM tests are expensive to run and should only be used when really necessary.

Maybe use `lowPrio`? Or add a `meta` that will build it only locally?

File: nixos/modules/services/misc/libreddit.nix (hunk 5 above)
Comment (fabaff) [reply nixos/modules/services/misc/libreddit.nix]: Changed, thanks.

File: nixos/modules/services/misc/libreddit.nix [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+with lib;
+
+  let
+    cfg = config.services.libreddit;
+
+    args = concatStringsSep " " ([
+      "--port ${toString cfg.port}"
+      "--address ${cfg.address}"
+    ] ++ optional cfg.redirect "--redirect-https");
```
Comment (aanderse) [inline nixos/modules/services/misc/libreddit.nix:11]: This implies we need an ssl certificate. I don't see one. Can you please explain?

File: nixos/modules/services/misc/libreddit.nix [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Open ports in the firewall for the libreddit web interface";
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
```
Comment (aanderse) [inline nixos/modules/services/misc/libreddit.nix:48]: Were you planning on putting this behind a reverse proxy? If not I would suggest mentioning somewhere that the sysadmin will need to configure this themselves.

File: nixos/modules/services/misc/libreddit.nix [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    systemd.services.libreddit = {
+        description = "Private front-end for Reddit";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        serviceConfig = {
+          DynamicUser = true;
+          ExecStart = "${pkgs.libreddit}/bin/libreddit ${args}";
```
Comment (aanderse) [inline nixos/modules/services/misc/libreddit.nix:55]: This program runs 100% stateless? Nice.

File: nixos/modules/services/misc/libreddit.nix (hunk 7 above)
Comment (fabaff) [reply nixos/modules/services/misc/libreddit.nix:11]: This feature was removed.

File: nixos/modules/services/misc/libreddit.nix (hunk 8 above)
Comment (fabaff) [reply nixos/modules/services/misc/libreddit.nix:48]: Ok, I will add this detail.

File: nixos/tests/libreddit.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (lukegb) [reply nixos/tests/libreddit.nix:11]: I think the VM test is fine for this case: this is testing both the package and the associated NixOS module.

File: nixos/modules/services/misc/libreddit.nix (hunk 7 above)
Comment (lukegb) [reply nixos/modules/services/misc/libreddit.nix:11]: Which feature? The SSL redirect feature still seems to be in your latest PR version - were you referring to a different feature?

**Comment by erdnaxe** [comment]
Libreddit is now shipping a `cargo.lock`. We don't need the patch to add the lock file anymore if this pull request is updated to latest libreddit version.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 113185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libreddit</li>
  </ul>
</details>

**Comment by aanderse** [comment]
@siraben a little bit premature on clicking merge in my opinion. There were a few points unresolved and I believe changes were requested on this PR. What's up?

**Comment by aanderse** [comment]
@fabaff are you able to address the points mentioned and continue the work here in a new PR? Let me know if you need a hand with anything.

**Comment by fabaff** [comment]
@aanderse, now as it's merged I have to fix the remaining issues :wink:

**Comment by siraben** [comment]
> @siraben a little bit premature on clicking merge in my opinion. There were a few points unresolved and I believe changes were requested on this PR. What's up?

Thanks, I'll be more thorough next time around.

**Comment by aanderse** [comment]
No problem @siraben. I appreciate the huge amount of work you always do on nixpkgs!! :heart:

**Comment by aanderse** [comment]
> @aanderse, now as it's merged I have to fix the remaining issues wink

Sounds good. As mentioned, let me know if I can help.

File: nixos/tests/libreddit.nix [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+  testScript = ''
+    machine.wait_for_unit("libreddit.service")
+    machine.wait_for_open_port("8080")
+    # The service wants to get data from https://www.reddit.com
+    machine.succeed("curl http://localhost:8080/")
```
[the remark this answers is not in the document]
Comment (erdnaxe) [reply nixos/tests/libreddit.nix:17]: I have proposed the change here: https://github.com/NixOS/nixpkgs/pull/133771

File: pkgs/servers/libreddit/default.nix [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+, stdenv
+, nixosTests
+, rustPlatform
+, fetchFromGitHub
+, Security
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "libreddit";
+  version = "0.10.1";
```
[the remark this answers is not in the document]
Comment (erdnaxe) [reply pkgs/servers/libreddit/default.nix:11]: I have proposed the change here: https://github.com/NixOS/nixpkgs/pull/133773

**Pull request merged**